ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryLedgerReaderWriter.scala

```diff
@@ -53,39 +49,58 @@ final class InMemoryLedgerReaderWriter(
     extends LedgerWriter
     with LedgerReader {
 
-  private val engine = Engine()
-
   private val currentState = new InMemoryState()
 
-  override def commit(correlationId: String, envelope: Array[Byte]): Future[SubmissionResult] =
-    Future {
-      val submission = Envelope
-        .openSubmission(envelope)
-        .getOrElse(throw new IllegalArgumentException("Not a valid submission in envelope"))
-      currentState.synchronized {
-        val stateInputs: Map[DamlStateKey, Option[DamlStateValue]] =
-          submission.getInputDamlStateList.asScala
-            .map(key => key -> currentState.state.get(key.toByteString))(breakOut)
-        val entryId = sequentialLogEntryId.next()
-        val (logEntry, damlStateUpdates) =
-          KeyValueCommitting.processSubmission(
-            engine,
-            entryId,
-            currentRecordTime(),
-            LedgerReader.DefaultConfiguration,
-            submission,
-            participantId,
-            stateInputs,
-          )
-        val stateUpdates = damlStateUpdates.map {
-          case (damlStateKey, value) => damlStateKey.toByteString -> value
+  private class InMemoryLedgerStateAccess(theParticipantId: ParticipantId)
+      extends LedgerStateAccess {
+    override def inTransaction[T](body: LedgerStateOperations => Future[T]): Future[T] =
+      Future {
+        currentState.state.synchronized {
+          body(new InMemoryLedgerStateOperations)
         }
-        currentState.log += LogEntry(entryId, Envelope.enclose(logEntry).toByteArray)
-        currentState.state ++= stateUpdates
-        dispatcher.signalNewHead(currentState.log.size)
+      }.flatten
+
+    override def participantId: String = theParticipantId
+  }
+
+  private class InMemoryLedgerStateOperations extends BatchingLedgerStateOperations {
```

Comment:
This can probably be an `object`; there's no state.

Reply:
Yes -- changed accordingly.